sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,641 @@
+				}
+			}
+
+		}
+
+		class HadoopInputFormatReader<T extends Object> extends BoundedSource.BoundedReader<KV<K, V>> {
+
+			private InputFormat<?,?> inputFormatObj;
+			private TaskAttemptContextImpl attemptContext;
+			private List<InputSplit> splits;
+			private ListIterator<InputSplit> splitsIterator;
+			private  RecordReader<T,T>  currentReader;
+			private KV<K, V> currentPair;
+			private volatile boolean done = false;
+			private final HadoopInputFormatBoundedSource<K, V> source;
+			private final SimpleFunction< T, K> simpleFuncForKeyTranslation;
+			private final SimpleFunction< T, V> simpleFuncForValueTranslation;
+			private long recordsReturned = 0;
+
+
+			public HadoopInputFormatReader(HadoopInputFormatBoundedSource<K,V> source, SimpleFunction keyTranslation,
+					SimpleFunction valueTranslation, InputSplit split) throws IOException , InstantiationException , IllegalAccessException , ClassNotFoundException  {
+				this.source = source;
+				Job job =  Job.getInstance(source.getConfiguration().getConfiguration());
+				inputFormatObj = (InputFormat<?,?>) job.getInputFormatClass().newInstance();
+				if (split != null) {
+					this.splits = ImmutableList.of(split);
+					this.splitsIterator = splits.listIterator();
+				}
+				this.simpleFuncForKeyTranslation = keyTranslation;
+				this.simpleFuncForValueTranslation = valueTranslation;
+			}
+
+			@Override
+			public HadoopInputFormatBoundedSource<K,V> getCurrentSource() {
+				return source;
+			}
+
+			@Override
+			public boolean start() throws IOException {
+				attemptContext = new TaskAttemptContextImpl(source.getConfiguration().getConfiguration(),
+						new TaskAttemptID());
+				InputSplit nextSplit = splitsIterator.next();
+				try {
+					@SuppressWarnings("unchecked")
+					RecordReader<T, T> reader = (RecordReader<T, T>) inputFormatObj.createRecordReader(nextSplit,attemptContext);
+					if (currentReader != null) {
+						currentReader.close();
+					}
+					currentReader = reader;
+					currentReader.initialize(nextSplit, attemptContext);
+					if (currentReader.nextKeyValue()) {
+						currentPair = nextPair();
+						return true;
+					}
+					currentReader.close();
+					currentReader = null;
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+					LOG.error(e.getMessage());
+					throw new IOException(e);
+				}
+				currentPair = null;
+				done = true;
+				return false;
+			}
+
+			@Override
+			public boolean advance() throws IOException {
+				try {
+					if (currentReader != null && currentReader.nextKeyValue()) {
+						currentPair = nextPair();
+						return true;
+					}
+					currentPair = null;
+					done = true;
+					return false;
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+					LOG.error(e.getMessage());
+					throw new IOException(e);
+				}
+			}
+
+
+			public KV<K, V> nextPair() throws IOException, InterruptedException {
```

Comment:
I think I mentioned the idea of extracting the code for key & value out - I believe you could have a single type parameterized method that handles both.